### ISSUE TITLE
HotString and Send unicode issue

### ISSUE THREAD

**Issue by vitia** [Lexikos/AutoHotkey_L#41]
Tested on 1.1.22.02 and 2.0-a068-5b3205b of AutoHotkey_L Unicode 64-bit.
HotString has no unicode symbols.
Send sends unicode symbol.
```
Test(c) {
	SendLevel 1 
	Send a%c%z
	;Send a%c%{Backspace}z
}

#InputLevel 1
1::Test("$")
2::Test("{U+0024}") ; $
3::Test("€")
4::Test("{U+20AC}") ; €

#InputLevel 0
::az::x
```
Input (without backspace): {Space}1{Space}2{Space}3{Space}4{Space}
Result:  a$z ax ax ax 
Expected: a$z a$z a€z a€z 

Input (with backspace): {Space}1{Space}2{Space}3{Space}4{Space}
Result:  x az az az 
Expected: x x x x

**Comment by joedf** [comment]
hmm.... I just tested your script.
I pressed `1234` and got `a$za$za€za€z`.

**Comment by vitia** [comment]
You need to press {Space} as well.
Results in previous post are for 04090409 US keybord layout.

There are results using F0010409 United States-International keybord layout:
Result (without backspace): a$z ax a€z ax 
Result (with backspace):  x az x az

**Comment by joedf** [comment]
hmmm I see, it looks like autohotkey is auto-triggering itself.
I know for hotkeys, we use `$`, I don't know what for hotstrings..

**Comment by vitia** [comment]
There is one more test.
It seems that hotstring engine incorrectly assumes that unicode symbol is a word boundary.
```
:*?:ss::š
::aassdd::xxyyzz
::dd::ww
```
Input: aassdd{Space}
Result: aašww 
Expected: aašdd

**Comment by vitia** [comment]
Actually bug shown in last post can be reproduced with non-Unicode characters.
```
:*?:ss::R
::aassdd::xxyyzz
::dd::ww
```
Input: aassdd{Space}
Result: aaRww 
Expected: aaRdd

**Comment by joedf** [comment]
I'm not sure if it's a bug or it's just buggy usage.... :wink:

**Comment by vitia** [comment]
100% bug, however Microsoft would call it a "feature" :)

**Comment by Lexikos** [comment]
**Sub-issue no. 1: hotstring recognizer ignoring special characters**

The Input command and the hotstring recognizer currently only recognize input generated by *keystrokes* or key combinations.  `{U+0024}` and `{U+20AC}` instruct the Send command to send a *character code* - this is done via the OS-level SendInput() function with specific parameters which result in a "VK_PACKET" message.  These are not handled by the hotstring recognizer, but probably could be.

`€` also does not correspond to a keystroke with your keyboard layout.  Since you are using a Unicode version of AutoHotkey, it is also sent as a "VK_PACKET" message.  On ANSI builds, characters which do not correspond to keystrokes are sent as Alt+Numpad combinations, which are also not handled by the hotstring recognizer.

> joedf: hmmm I see, it looks like autohotkey is auto-triggering itself.

Yes, that was the purpose of setting a `SendLevel` greater than the `#InputLevel` of the hotstring.  With default settings, Send will not trigger the hotstring.  This is not the problem.  The problem is that the hotstring recognizer is only recognizing some of the characters which are being sent.

**Sub-issue no. 2: auto-replace hotstring acting as a word boundary**

> vitia: 100% bug

Obviously if *your usage* was at fault, you would be the wrong person to decide that it's a bug.  Arguing that it's a bug won't get it fixed any faster.

That said, I agree it *seems* like a bug.  However, it appears to be intentional:

````C++
if (*hs.mReplacement)
{
	// Since the buffer no longer reflects what is actually on screen to the left
	// of the caret position (since a replacement is about to be done), reset the
	// buffer, except for any end-char (since that might legitimately form part
	// of another hot string adjacent to the one just typed).
````
(These comments were written by Chris, the original developer of AutoHotkey.)

To produce the behaviour *you* expect, auto-replace hotstrings would need to adjust the hotstring buffer to reflect the changes that the hotstring makes.  However, auto-replace hotstrings are documented to operate at SendLevel 0; i.e. they are supposed to be **ignored** by the hotstring recognizer.

It would arguably be correct for the `aassdd` hotstring to fire - since the backspace and R keystrokes are supposed to be ignored - even though it would produce the wrong result due to the hotstring changing the number of visible characters.

> [SendLevel](http://ahkscript.org/docs/commands/SendLevel.htm): Auto-replace hotstrings always generate keystrokes at level 0, since it is usually undesirable for the replacement text to trigger another hotstring or hotkey. To work around this, use a non-auto-replace hotstring and the Send command.

If you want to work around the issue, use a non-auto-replace hotstring:
````autohotkey
:?*:ss::
	SendLevel 1
	Send R
	return
::dd::ww
````

**Comment by vitia** [comment]
Your explanation is excellent. 
That's what I ended up with.

```
::aassdd::xxyyzz
::dd::ww
::ad::xz
:*?:ss::
	SendLevel 1
	Send š
	return
```
Input: adss{Space}
Result: xz 
Expected: adš 

However now with your explanation I understand why this happens:
```
input: adss
	| input  | screen | recognizer |
	| "adss" | "adss" | "adss"     |
`:*?:ss::Send š` hotstring kicks in and sends {Backspace 2} and š as VK_PACKET, however hotstring recognizer gets backspaces, however ignores VK_PACKET
	| input  | screen | recognizer |
	| "adss" | "adš"  | "ad"       |

input: {Space}
	| input   | screen | recognizer |
	| "adss " | "ad "  | "ad "      |
`::ad::xz` hotstring kicks in and sends "{Backspace 3}xz "
	| adss "  | "xz "  | "xz "      |
```
I came to this hack to trick hotstring recognizer: `Send 0{U+0008}`
This works for me as I have no numbers in my hotstrings.

```
SetKeyDelay -1
::aassdd::xxyyzz
::dd::ww
::ad::xz
:*?:ss::
	SendLevel 1
	Send 0{U+0008}
	Send š
	return
```
Input: adss{Space}
Result: adš 
Input: aassdd{Space}
Result: aašdd 
Input: ssdd{Space}
Result: šdd  

All results are as expected.
I hope I can get rid of this hack some day, because "0" does flicker sometimes.

Thank you very much :)

**Comment by Lexikos** [comment]
Sub-issue no. 1 was fixed by commit 5ba733b5 (v1.1.22.06).
Sub-issue no. 2 won't be fixed.
